internal/defence/limiter.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package defence
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"storj.io/storj/internal/sync2"
+)
+
+// Attacker stores information about attacker entity
+type Attacker struct {
+	limiter *rate.Limiter
+	Expire  time.Time
+	// TODO: could be interface{}
+	key string
+
+	IsBanned bool
+}
+
+func (attacker *Attacker) attack(attemptTime time.Time, banDuration time.Duration) bool {
+	// is not cleared by Clear method of Limiter
+	if attacker.IsBanned && attemptTime.After(attacker.Expire) {
+		attacker.IsBanned = false
+	}
+
+	if attacker.IsBanned {
+		return true
+	}
+
+	isBanned := !attacker.limiter.AllowN(attemptTime, 1)
+
+	if isBanned {
+		attacker.IsBanned = true
+		attacker.Expire = attemptTime.Add(banDuration)
+	}
+
+	return isBanned
+}
+
+// Limiter is used to store and manage list of banned entities
+type Limiter struct {
```

Comment:
The `banned` and `attacker` seem bad word choices. e.g. `RateLimited` would sound nicer

Reply:
there still are things naming attackers/banning